File: infrastructures/ddd/python/src/api/controllers/example_controller.py
"""
Example controller demonstrating controller layer patterns.
Controllers:
- Live in /api/controllers/
- Have Controller suffix
- Handle HTTP only: bind, validate, authorize, map DTOs
- NO business logic (delegate to services)
- Use BaseController for common functionality

Middleware order (automatically applied):
1. Correlation ID → 2. Logging → 3. Authentication → 
4. Authorization/Ownership → 5. UnitOfWork → 6. Controller →
7. UnitOfWork commit/rollback → 8. Error Handling
"""

from typing import List
from api.controllers.infra import BaseController
from api.dto.requests.create_example_request import CreateExampleRequest
from api.dto.requests.update_example_request import UpdateExampleRequest
from api.dto.responses.example_response import ExampleResponse
from application.services.i_example_service import IExampleService


class ExampleController(BaseController):
    """
    Example controller demonstrating REST API patterns.
    
    This controller would be decorated with framework-specific
    route decorators in a real application:
    
    FastAPI:
    ```python
    from fastapi import APIRouter, Depends, HTTPException, status
    
    router = APIRouter(prefix="/api/v1/examples", tags=["examples"])
    
    @router.post("/", status_code=status.HTTP_201_CREATED)
    async def create_example(
        request: CreateExampleRequest,
        service: IExampleService = Depends()
    ) -> str:
        controller = ExampleController(service)
        return await controller.create_example(request)
    ```
    
    Flask:
    ```python
    from flask import Blueprint, request, jsonify
    
    bp = Blueprint('examples', __name__, url_prefix='/api/v1/examples')
    
    @bp.route('/', methods=['POST'])
    async def create_example():
        req = CreateExampleRequest(**request.json)
        controller = ExampleController(example_service)
        example_id = await controller.create_example(req)
        return jsonify(example_id), 201
    ```
    """
    
    def __init__(self, p_example_service: IExampleService):
        """
        Initialize the example controller.
        
        Args:
            p_example_service: The example service
        """
        super().__init__()
        self._example_service = p_example_service
    
    async def create_example(
        self,
        p_request: CreateExampleRequest
    ) -> str:
        """
        Creates a new example.
        POST /api/v1/examples
        
        Args:
            p_request: The create request
            
        Returns:
            The ID of the created example
        """
        # Validate request (edge validation)
        self.validate_request(p_request)
        
        # Get user ID from auth context
        user_id = self.get_user_id()
        correlation_id = self.get_correlation_id()
        
        # Delegate to service (business logic happens here)
        example_id = await self._example_service.create_example_async(
            p_request,
            user_id,
            correlation_id
        )
        
        return example_id
    
    async def get_example(self, p_id: str) -> ExampleResponse:
        """
        Gets an example by ID.
        GET /api/v1/examples/{id}
        
        Args:
            p_id: The example ID
            
        Returns:
            The example response
        """
        return await self._example_service.get_example_async(p_id)
    
    async def list_examples(
        self,
        p_skip: int = 0,
        p_take: int = 10
    ) -> List[ExampleResponse]:
        """
        Lists examples with pagination.
        GET /api/v1/examples?skip=0&take=10
        
        Args:
            p_skip: Number of records to skip
            p_take: Number of records to take
            
        Returns:
            List of example responses
        """
        return await self._example_service.list_examples_async(p_skip, p_take)
    
    async def update_example(
        self,
        p_id: str,
        p_request: UpdateExampleRequest
    ) -> None:
        """
        Updates an example.
        PUT /api/v1/examples/{id}
        
        Args:
            p_id: The example ID
            p_request: The update request
        """
        # Validate request
        self.validate_request(p_request)
        
        # Get user ID
        user_id = self.get_user_id()
        
        # Delegate to service
        await self._example_service.update_example_async(
            p_id,
            p_request,
            user_id
        )
    
    async def delete_example(self, p_id: str) -> None:
        """
        Deletes an example.
        DELETE /api/v1/examples/{id}
        
        Args:
            p_id: The example ID
        """
        user_id = self.get_user_id()
        await self._example_service.delete_example_async(p_id, user_id)
    
    async def activate_example(self, p_id: str) -> None:
        """
        Activates an example.
        POST /api/v1/examples/{id}/activate
        
        Args:
            p_id: The example ID
        """
        user_id = self.get_user_id()
        await self._example_service.activate_example_async(p_id, user_id)
    
    async def deactivate_example(self, p_id: str) -> None:
        """
        Deactivates an example.
        POST /api/v1/examples/{id}/deactivate
        
        Args:
            p_id: The example ID
        """
        user_id = self.get_user_id()
        await self._example_service.deactivate_example_async(p_id, user_id)

File: infrastructures/ddd/typescript/src/api/controllers/ExampleController.ts
import { Request, Response, NextFunction } from 'express';
import { BaseController } from './infra/BaseController';
import { IExampleService } from '../../application/services/IExampleService';
import { CreateExampleRequest } from '../dto/v1/requests/CreateExampleRequest';
import { UpdateExampleRequest } from '../dto/v1/requests/UpdateExampleRequest';

/**
 * Example controller demonstrating controller layer patterns.
 * Controllers:
 * - Live in /api/controllers/
 * - Have Controller suffix
 * - Handle HTTP only: bind, validate, authorize, map DTOs
 * - NO business logic (delegate to services)
 * - Are concrete implementations (no /impl/ subdirectory)
 * 
 * Middleware order (automatically applied):
 * 1. Correlation ID → 2. Logging → 3. Authentication → 
 * 4. Authorization/Ownership → 5. UnitOfWork → 6. Controller →
 * 7. UnitOfWork commit/rollback → 8. Error Handling
 * 
 * Routes (example with Express):
 * POST   /api/v1/example
 * GET    /api/v1/example/:id
 * GET    /api/v1/example
 * PUT    /api/v1/example/:id
 * DELETE /api/v1/example/:id
 * POST   /api/v1/example/:id/activate
 * POST   /api/v1/example/:id/deactivate
 */
export class ExampleController extends BaseController {
  private readonly _exampleService: IExampleService;

  constructor(p_exampleService: IExampleService) {
    super();
    this._exampleService = p_exampleService;
  }

  /**
   * Creates a new example.
   * POST /api/v1/example
   */
  public async CreateExample(
    p_request: Request,
    p_response: Response,
    p_next: NextFunction
  ): Promise<void> {
    try {
      // Validate request (edge validation)
      const body = p_request.body as CreateExampleRequest;
      this.ValidateRequest(body);

      // Get user ID from auth context
      const userId = this.GetUserId(p_request);
      const correlationId = this.GetCorrelationId(p_request);

      // Delegate to service (business logic happens here)
      const exampleId = await this._exampleService.CreateExampleAsync(body, userId, correlationId);

      // Return created response
      p_response.status(201).json({ id: exampleId });
    } catch (error) {
      p_next(error);
    }
  }

  /**
   * Gets an example by ID.
   * GET /api/v1/example/:id
   */
  public async GetExample(
    p_request: Request,
    p_response: Response,
    p_next: NextFunction
  ): Promise<void> {
    try {
      const id = p_request.params.id;
      const result = await this._exampleService.GetExampleAsync(id);
      p_response.status(200).json(result);
    } catch (error) {
      p_next(error);
    }
  }

  /**
   * Lists examples with pagination.
   * GET /api/v1/example?skip=0&take=10
   */
  public async ListExamples(
    p_request: Request,
    p_response: Response,
    p_next: NextFunction
  ): Promise<void> {
    try {
      const skip = parseInt(p_request.query.skip as string) || 0;
      const take = parseInt(p_request.query.take as string) || 10;
      
      const results = await this._exampleService.ListExamplesAsync(skip, take);
      p_response.status(200).json(results);
    } catch (error) {
      p_next(error);
    }
  }

  /**
   * Updates an example.
   * PUT /api/v1/example/:id
   */
  public async UpdateExample(
    p_request: Request,
    p_response: Response,
    p_next: NextFunction
  ): Promise<void> {
    try {
      const id = p_request.params.id;
      const body = p_request.body as UpdateExampleRequest;
      this.ValidateRequest(body);

      const userId = this.GetUserId(p_request);
      await this._exampleService.UpdateExampleAsync(id, body, userId);

      p_response.status(204).send();
    } catch (error) {
      p_next(error);
    }
  }

  /**
   * Deletes an example.
   * DELETE /api/v1/example/:id
   */
  public async DeleteExample(
    p_request: Request,
    p_response: Response,
    p_next: NextFunction
  ): Promise<void> {
    try {
      const id = p_request.params.id;
      const userId = this.GetUserId(p_request);
      
      await this._exampleService.DeleteExampleAsync(id, userId);
      p_response.status(204).send();
    } catch (error) {
      p_next(error);
    }
  }

  /**
   * Activates an example.
   * POST /api/v1/example/:id/activate
   */
  public async ActivateExample(
    p_request: Request,
    p_response: Response,
    p_next: NextFunction
  ): Promise<void> {
    try {
      const id = p_request.params.id;
      const userId = this.GetUserId(p_request);
      
      await this._exampleService.ActivateExampleAsync(id, userId);
      p_response.status(204).send();
    } catch (error) {
      p_next(error);
    }
  }

  /**
   * Deactivates an example.
   * POST /api/v1/example/:id/deactivate
   */
  public async DeactivateExample(
    p_request: Request,
    p_response: Response,
    p_next: NextFunction
  ): Promise<void> {
    try {
      const id = p_request.params.id;
      const userId = this.GetUserId(p_request);
      
      await this._exampleService.DeactivateExampleAsync(id, userId);
      p_response.status(204).send();
    } catch (error) {
      p_next(error);
    }
  }
}
